### PR TITLE
Streamline the validation of inputs during submission to a single place

### PR DESCRIPTION
**Pull request by geoffjentry**

### CODE REVIEW

File: src/main/scala/cromwell/engine/WorkflowDescriptor.scala [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+    WorkflowOptions.fromJsonString(optionsJson) match {
+      case Success(o) => buildValidateOptions(o, id, backend)
+      case Failure(e) => s"Workflow ${id.toString} contains bad options JSON: ${e.getLocalizedMessage}".failureNel
+    }
+  }
+
+  private def buildValidateOptions(options: WorkflowOptions, id: WorkflowId, backend: Backend): ErrorOr[WorkflowOptions] = {
+    val validatedOptions = options.successNel[String]
+    val backendOptionsDisjunction = for {
+      o <- validatedOptions.disjunction
```
Comment (mcovarr) [inline src/main/scala/cromwell/engine/WorkflowDescriptor.scala]: Why the dance with `options.successNel[String]`?  Can't you just pass `options` to `validateBackendOptions(...)`?

File: src/main/scala/cromwell/engine/WorkflowDescriptor.scala [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+      b <- validateBackendOptions(id, o, backend).disjunction
+    } yield b
+    val backendOptions = backendOptionsDisjunction.validation
+    (validatedOptions |@| backendOptions) { (options, b) => options }
```
Comment (mcovarr) [inline src/main/scala/cromwell/engine/WorkflowDescriptor.scala]: `b` -> `_`

File: src/main/scala/cromwell/engine/WorkflowDescriptor.scala [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+        validatedDescriptor.validation
+      case scalaz.Failure(f) => f.list.mkString("").failureNel
+    }
+  }
+
+  private def validateNamespace(source: WdlSource, backend: Backend): ErrorOr[NamespaceWithWorkflow] = {
+    try {
+      NamespaceWithWorkflow.load(source, backend.backendType).successNel
+    } catch {
+      case e: Exception => s"Workflow ${id.toString} unable to load namespace: ${e.getLocalizedMessage}".failureNel
```
Comment (mcovarr) [inline src/main/scala/cromwell/engine/WorkflowDescriptor.scala]: If there are errors it would probably be better to only output the name of the workflow once rather than for every error.

File: src/main/scala/cromwell/engine/WorkflowDescriptor.scala [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+  def getConfigOption(conf: Config, key: String): Option[Config] = if (conf.hasPath(key)) Option(conf.getConfig(key)) else None
+
+  def configCallCaching(conf: Config) = lookupWithDefault(conf, "call-caching", "enabled", DefaultCallCachingValue)
+  def lookupDockerHash(conf: Config) = lookupWithDefault(conf, "call-caching", "lookup-docker-hash", DefaultLookupDockerHash)
+  def lookupWithDefault(conf: Config, stanza: String, key: String, default: Boolean) = {
+    (for {
+      config <- getConfigOption(conf, stanza)
+      value <- config.getBooleanOption(key)
+    } yield value) getOrElse default
+  }
```
Comment (mcovarr) [inline src/main/scala/cromwell/engine/WorkflowDescriptor.scala]: sweet

File: src/main/scala/cromwell/engine/WorkflowDescriptor.scala [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+      case _ => s"Workflow ${id.toString} contains bad inputs JSON: $json".failureNel
+    }
+  }
+
+  private def validateCoercedInputs(id: WorkflowId,
+                                    rawInputs: Map[String, JsValue],
+                                    namespace: NamespaceWithWorkflow): ErrorOr[WorkflowCoercedInputs] = {
+    namespace.coerceRawInputs(rawInputs) match {
+      case Success(r) => r.successNel
+      case Failure(e) => e.getLocalizedMessage.failureNel
```
Comment (mcovarr) [inline src/main/scala/cromwell/engine/WorkflowDescriptor.scala]: localized message?

File: src/main/scala/cromwell/engine/WorkflowDescriptor.scala (hunk 3 above)
Comment (geoffjentry) [inline src/main/scala/cromwell/engine/WorkflowDescriptor.scala]: Speaking of redundancies

File: src/main/scala/cromwell/engine/WorkflowDescriptor.scala [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+    case Some(x) if x.toUpperCase.contains("SERVER") => makeFileLogger(
+      Paths.get(props.getOrElse("LOG_ROOT", ".")),
+      s"workflow.$id.log",
+      Level.toLevel(props.getOrElse("LOG_LEVEL", "debug"))
+    )
+    case _ => NOPLogger.NOP_LOGGER
+  }
+
+
+
```
Comment (mcovarr) [inline src/main/scala/cromwell/engine/WorkflowDescriptor.scala]: well I guess if you take out a bunch of blank lines you need to put new ones back in... :stuck_out_tongue:

File: src/main/scala/cromwell/engine/WorkflowDescriptor.scala [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+  private def buildValidateOptions(options: WorkflowOptions, id: WorkflowId, backend: Backend): ErrorOr[WorkflowOptions] = {
+    val validatedOptions = options.successNel[String]
+    val backendOptionsDisjunction = for {
+      o <- validatedOptions.disjunction
+      b <- validateBackendOptions(id, o, backend).disjunction
+    } yield b
+    val backendOptions = backendOptionsDisjunction.validation
+    (validatedOptions |@| backendOptions) { (options, b) => options }
+  }
```
Comment (mcovarr) [inline src/main/scala/cromwell/engine/WorkflowDescriptor.scala]: I think this is equivalent

```scala
  private def buildValidateOptions(options: WorkflowOptions, id: WorkflowId, backend: Backend): ErrorOr[WorkflowOptions] = {
    validateBackendOptions(id, options, backend) map { _ => options }
  }
```

Comment (mcovarr) [inline src/main/scala/cromwell/engine/WorkflowDescriptor.scala]: or just have `validateBackendOptions` return `ErrorOr[WorkflowOptions]` instead of `Error[Unit]` and this method can go away

File: src/test/scala/cromwell/webservice/CromwellApiServiceSpec.scala [hunk 8]
```diff
@@ -357,12 +351,9 @@ class CromwellApiServiceSpec extends FlatSpec with CromwellApiService with Scala
...
       submitRoute ~>
       check {
-        assertResult("Expecting JSON object for workflowInputs and workflowOptions fields") {
-          responseAs[String]
-        }
-        assertResult(StatusCodes.BadRequest) {
-          status
-        }
+        assertResult(StatusCodes.BadRequest) { status }
+        val x = responseAs[String]
```
Comment (mcovarr) [inline src/test/scala/cromwell/webservice/CromwellApiServiceSpec.scala]: stray `x` disorder appears to be contagious

File: src/main/scala/cromwell/engine/WorkflowDescriptor.scala [hunk 9]
```diff
@@ -0,0 +1,205 @@
+package cromwell.engine
```
Comment (mcovarr) [inline src/main/scala/cromwell/engine/WorkflowDescriptor.scala]: good call pulling this into its own file

File: src/test/scala/cromwell/webservice/CromwellApiServiceSpec.scala [hunk 10]
```diff
@@ -51,8 +52,12 @@ class MockWorkflowManagerActor extends Actor  {
 
   def receive = {
     case SubmitWorkflow(sources) =>
-      sender ! MockWorkflowManagerActor.submittedWorkflowId
-
+      val id = MockWorkflowManagerActor.submittedWorkflowId
+      val response = Try(WorkflowDescriptor(id, sources)) match {
```
Comment (mcovarr) [inline src/test/scala/cromwell/webservice/CromwellApiServiceSpec.scala]: `Future.fromTry`

**Comment by mcovarr** [comment]
:+1:

File: src/main/scala/cromwell/engine/WorkflowDescriptor.scala (hunk 6 above)
Comment (geoffjentry) [inline src/main/scala/cromwell/engine/WorkflowDescriptor.scala]: Heh. I need to leave you all *some* whitespace

File: src/main/scala/cromwell/engine/WorkflowDescriptor.scala (hunk 1 above)
Comment (geoffjentry) [inline src/main/scala/cromwell/engine/WorkflowDescriptor.scala]: Pretty sure that evolved a bit and I didn't notice I'd perturbed the thing to the extent I had

File: src/main/scala/cromwell/engine/WorkflowDescriptor.scala (hunk 5 above)
Comment (geoffjentry) [inline src/main/scala/cromwell/engine/WorkflowDescriptor.scala]: I honestly don't know why I used that, and it's throughout. Either something I started from here used it or something I cribbed off of at some point used it. 

I don't see the need to use it anywhere, do you?

File: src/main/scala/cromwell/engine/WorkflowDescriptor.scala (hunk 4 above)
Comment (geoffjentry) [inline src/main/scala/cromwell/engine/WorkflowDescriptor.scala]: I was just thinking that this could move over to ConfigUtil (speaking of, wth is that validation pimped class over there?)

File: src/test/scala/cromwell/webservice/CromwellApiServiceSpec.scala (hunk 8 above)
Comment (geoffjentry) [inline src/test/scala/cromwell/webservice/CromwellApiServiceSpec.scala]: When I grow up I want to be just like @cjllanwarne

File: src/test/scala/cromwell/webservice/CromwellApiServiceSpec.scala (hunk 10 above)
Comment (geoffjentry) [inline src/test/scala/cromwell/webservice/CromwellApiServiceSpec.scala]: That actually wasn't a fromTry - in the success case it was passing id back and in the failure it was passing the try's failure

File: src/test/scala/cromwell/webservice/CromwellApiServiceSpec.scala [hunk 11]
```diff
@@ -51,7 +52,18 @@ class MockWorkflowManagerActor extends Actor  {
 
   def receive = {
     case SubmitWorkflow(sources) =>
-      sender ! MockWorkflowManagerActor.submittedWorkflowId
+      val id = MockWorkflowManagerActor.submittedWorkflowId
+      /*
```
Comment (geoffjentry) [inline src/test/scala/cromwell/webservice/CromwellApiServiceSpec.scala]: @mcovarr I resurrected this w/ comment - it's not a simple fromTry. Do you see a better path here?

File: src/main/scala/cromwell/engine/backend/local/SharedFileSystem.scala [hunk 12]
```diff
@@ -17,7 +17,6 @@ import cromwell.engine.backend._
 import cromwell.engine.workflow.{CallKey, WorkflowOptions}
 import cromwell.util.TryUtil
 import org.apache.commons.io.FileUtils
-
```
Comment (kshakir) [inline src/main/scala/cromwell/engine/backend/local/SharedFileSystem.scala]: IntelliJ, by default, likes this separate line.

File: src/test/scala/cromwell/logging/WorkflowLoggerSpec.scala [hunk 13]
```diff
@@ -6,7 +6,7 @@ import cromwell.CromwellTestkitSpec
 import cromwell.binding.values.WdlValue
 import cromwell.engine.backend.local.{LocalBackend, LocalBackendCall}
 import cromwell.engine.workflow.CallKey
-import cromwell.engine.{AbortRegistrationFunction, WorkflowDescriptor, WorkflowId, WorkflowSourceFiles}
+import cromwell.engine.{WorkflowDescriptor, AbortRegistrationFunction, WorkflowId, WorkflowSourceFiles}
```
Comment (kshakir) [inline src/test/scala/cromwell/logging/WorkflowLoggerSpec.scala]: Revert this change

File: src/test/scala/cromwell/webservice/CromwellApiServiceSpec.scala [hunk 14]
```diff
@@ -344,12 +349,8 @@ class CromwellApiServiceSpec extends FlatSpec with CromwellApiService with Scala
     Post("/workflows/$version", FormData(Seq("wdlSource" -> HelloWorld.wdlSource(), "workflowInputs" -> CromwellApiServiceSpec.MalformedInputsJson))) ~>
       submitRoute ~>
       check {
-        assertResult("Expecting JSON object for workflowInputs and workflowOptions fields") {
-          responseAs[String]
-        }
-        assertResult(StatusCodes.BadRequest) {
-          status
-        }
+        assertResult(StatusCodes.BadRequest) { status }
```
Comment (kshakir) [inline src/test/scala/cromwell/webservice/CromwellApiServiceSpec.scala]: Change these if you'd like, but FYI, IntelliJ's default auto formatter will restore the blocks to a separate line.

Comment (geoffjentry) [inline src/test/scala/cromwell/webservice/CromwellApiServiceSpec.scala]: IntelliJ is stupid :'(

File: src/main/scala/cromwell/engine/WorkflowDescriptor.scala (hunk 4 above)
Comment (geoffjentry) [inline src/main/scala/cromwell/engine/WorkflowDescriptor.scala]: I added some comments as when I started looking at doing what I was talking about doing, there were multiple cans of worms I saw opening up. Since this change itself was a can of worms which came out of a can of worms i decided to avoid future cans of worms for now.
5124hyepjal

Comment (kshakir) [inline src/main/scala/cromwell/engine/WorkflowDescriptor.scala]: `EnhancedValidation` is used over in our [JES code](https://github.com/broadinstitute/cromwell/blob/6b61833d088f224eda347617c8f251a182ac5175/src/main/scala/cromwell/engine/backend/jes/JesAttributes.scala#L55-L68).

Btw, `EnhancedValidation` in lenthall was copied from here: https://github.com/broadinstitute/cromwell/commit/8642363. Looking at the [history](https://github.com/broadinstitute/cromwell/commits/develop/src/main/scala/cromwell/engine/backend/jes/JesAttributes.scala) of `JesAttributes`, it looks like the cromwell `ConfigUtil` was brought back from the dead with the development / merging of #225.

File: src/test/scala/cromwell/webservice/CromwellApiServiceSpec.scala (hunk 11 above)
Comment (mcovarr) [inline src/test/scala/cromwell/webservice/CromwellApiServiceSpec.scala]:
```scala
      val response = Future.fromTry(Try(WorkflowDescriptor(id, sources)) map { _ => id }) 
      response pipeTo sender
```

File: src/test/scala/cromwell/webservice/CromwellApiServiceSpec.scala [hunk 15]
```diff
@@ -51,7 +52,18 @@ class MockWorkflowManagerActor extends Actor  {
 
   def receive = {
     case SubmitWorkflow(sources) =>
-      sender ! MockWorkflowManagerActor.submittedWorkflowId
+      val id = MockWorkflowManagerActor.submittedWorkflowId
+      /*
+        id will always succed, but WorkflowDescriptor might not. If all of this works we
```
Comment (mcovarr) [inline src/test/scala/cromwell/webservice/CromwellApiServiceSpec.scala]: succed -> succeed

File: src/main/scala/cromwell/engine/WorkflowDescriptor.scala [hunk 16]
```diff
@@ -0,0 +1,196 @@
...
+    }
+  }
+
+  private def validateBackendOptions(id: WorkflowId, options: WorkflowOptions, backend: Backend): ErrorOr[WorkflowOptions] = {
+    try {
+      backend.assertWorkflowOptions(options)
+      options.successNel
+    } catch {
+      case e: Exception =>
+        s"Workflow ${id.toString} has invalid options for backend ${backend.backendType}: ${e.getMessage}".failureNel
```
Comment (kshakir) [inline src/main/scala/cromwell/engine/WorkflowDescriptor.scala]: `${id.toString}` => `$id`

File: src/main/scala/cromwell/engine/WorkflowDescriptor.scala (hunk 4 above)
Comment (geoffjentry) [inline src/main/scala/cromwell/engine/WorkflowDescriptor.scala]: Right, I just meant "why is a validation thing in config". Not a big deal.

Your second part is tied to what I mean. There's a real mess involved there but things work as-is and considering I'm trying to back my way out of a deep rabbit hole I'm just going to leave it all mostly the way it is for now and look at it later

File: src/test/scala/cromwell/webservice/CromwellApiServiceSpec.scala (hunk 11 above)
Comment (geoffjentry) [inline src/test/scala/cromwell/webservice/CromwellApiServiceSpec.scala]: duh

File: src/main/scala/cromwell/engine/WorkflowDescriptor.scala (hunk 16 above)
Comment (geoffjentry) [inline src/main/scala/cromwell/engine/WorkflowDescriptor.scala]: While I believe you're correct, I'll point out WorkflowId.toString is used heavily throughout. I suspect perhaps at one point it mattered

**Comment by mcovarr** [comment]
Per some of @kshakir's comments, IntelliJ has a *default* formatting which we have not made any attempt to change.  If these defaults cause consternation (or code review comments), we might consider developing a group style which we could all share and configure our IntelliJ's to use that.  I definitely don't want to be fighting IntelliJ but I'm fine with moving away from the defaults if that's what the team wants.

File: src/main/scala/cromwell/engine/WorkflowDescriptor.scala (hunk 16 above)
Comment (geoffjentry) [inline src/main/scala/cromwell/engine/WorkflowDescriptor.scala]: Ah I see. I think instead it's just people getting confused. Outside of the string context it's just the identity, but inside interpolation it's not. I'll change it in the files I'm fiddling with

File: src/main/scala/cromwell/engine/WorkflowDescriptor.scala [hunk 17]
```diff
@@ -0,0 +1,196 @@
...
+    val rawInputs = validateRawInputs(id, sourceFiles.inputsJson)
+    val options = validateWorkflowOptions(id, sourceFiles.workflowOptionsJson, backend)
+    (namespace |@| rawInputs |@| options) { (_, _, _) } match {
+      case scalaz.Success((n, r, o)) =>
+        val validatedDescriptor = for {
+          c <- validateCoercedInputs(id, r, n).disjunction
+          d <- validateDeclarations(id, n, o, c, backend).disjunction
+        } yield WorkflowDescriptor(id, sourceFiles, o, r, n, c, d, backend, configCallCaching(conf), lookupDockerHash(conf))
+        validatedDescriptor.validation
+      case scalaz.Failure(f) => f.list.mkString("").failureNel
```
Comment (kshakir) [inline src/main/scala/cromwell/engine/WorkflowDescriptor.scala]: I doubt we have a test for multiple failures? Maybe in this PR, as this is copy paste, we don't need one. But I also doubt the failure strings are supposed to be concatenated without at least a `", "`.

Comment (geoffjentry) [inline src/main/scala/cromwell/engine/WorkflowDescriptor.scala]: That's a good catch because I know they are separated via newlines when they print out. Lemme look.

**Comment by kshakir** [comment]
:+1:

File: src/main/scala/cromwell/engine/WorkflowDescriptor.scala (hunk 17 above)
Comment (geoffjentry) [inline src/main/scala/cromwell/engine/WorkflowDescriptor.scala]: I'm glad you pointed this out @kshakir - the answer to the mystery was that this was working by magical happenstance

**Comment by geoffjentry** [comment]
Merging as the failing test is that same stupid integration test w/ dockerhub failing on develop

**Comment by mcovarr** [comment]
If that test is chronically failing due to Docker Hub flakiness we should tag it as Integration so it doesn't break our builds.

**Comment by mcovarr** [comment]
...and then we should mock out Docker Hub so we don't lose the value of the test, then re-enable.

**Comment by geoffjentry** [comment]
@mcovarr Yes. It's ridiculous that we get build failures due to external services failing.
